File: lib/rmhark-ai/__tests__/dm-policy.test.ts
import { describe, it, expect } from 'vitest';
import {
  needsReactiveReply,
  canBotMessage,
  decideInitiation,
  formatDmHistory,
} from '@/lib/rmhark-ai/dm-policy';

const BOT = 'bot-1';
const HUMAN = 'human-1';
const at = (ms: number) => ({ senderId: HUMAN, createdAt: new Date(ms) });

describe('needsReactiveReply', () => {
  it('is false for an empty conversation', () => {
    expect(needsReactiveReply([], BOT)).toBe(false);
  });
  it('is true when the human spoke last', () => {
    expect(
      needsReactiveReply(
        [{ senderId: BOT, createdAt: new Date(1) }, { senderId: HUMAN, createdAt: new Date(2) }],
        BOT,
      ),
    ).toBe(true);
  });
  it('is false when the bot already replied last', () => {
    expect(
      needsReactiveReply(
        [{ senderId: HUMAN, createdAt: new Date(1) }, { senderId: BOT, createdAt: new Date(2) }],
        BOT,
      ),
    ).toBe(false);
  });
});

describe('canBotMessage', () => {
  it('blocks NONE', () => {
    expect(canBotMessage({ dmPrivacy: 'NONE', humanFollowsBot: true })).toBe(false);
  });
  it('allows FOLLOWERS only when the human follows the bot', () => {
    expect(canBotMessage({ dmPrivacy: 'FOLLOWERS', humanFollowsBot: true })).toBe(true);
    expect(canBotMessage({ dmPrivacy: 'FOLLOWERS', humanFollowsBot: false })).toBe(false);
  });
  it('allows EVERYONE', () => {
    expect(canBotMessage({ dmPrivacy: 'EVERYONE', humanFollowsBot: false })).toBe(true);
  });
});

describe('decideInitiation', () => {
  const base = { botId: BOT, now: 1_000_000, followupSilenceMs: 1000 };
  it('opens when there is no conversation', () => {
    expect(decideInitiation({ ...base, messages: null })).toBe('opener');
    expect(decideInitiation({ ...base, messages: [] })).toBe('opener');
  });
  it('skips when the human has ever replied', () => {
    expect(
      decideInitiation({
        ...base,
        messages: [{ senderId: BOT, createdAt: new Date(0) }, at(5)],
      }),
    ).toBe('skip');
  });
  it('follows up once enough silence has passed after a lone opener', () => {
    expect(
      decideInitiation({ ...base, messages: [{ senderId: BOT, createdAt: new Date(0) }] }),
    ).toBe('followup');
  });
  it('skips a lone opener that is still within the silence window', () => {
    expect(
      decideInitiation({ ...base, messages: [{ senderId: BOT, createdAt: new Date(999_500) }] }),
    ).toBe('skip');
  });
  it('gives up after two unanswered bot messages', () => {
    expect(
      decideInitiation({
        ...base,
        messages: [
          { senderId: BOT, createdAt: new Date(0) },
          { senderId: BOT, createdAt: new Date(1) },
        ],
      }),
    ).toBe('skip');
  });
});

describe('formatDmHistory', () => {
  it('labels messages from the bot perspective, preserving order', () => {
    expect(
      formatDmHistory(
        [
          { senderId: HUMAN, content: 'hey' },
          { senderId: BOT, content: 'hi there' },
        ],
        BOT,
      ),
    ).toEqual([
      { from: 'them', text: 'hey' },
      { from: 'you', text: 'hi there' },
    ]);
  });
});

File: go-services/internal/botworker/dmpolicy_test.go
package botworker

import (
	"testing"
	"time"
)

func TestCanBotMessage(t *testing.T) {
	cases := []struct {
		privacy DmPrivacy
		follows bool
		want    bool
	}{
		{"NONE", true, false},
		{"NONE", false, false},
		{"FOLLOWERS", true, true},
		{"FOLLOWERS", false, false},
		{"EVERYONE", false, true},
		{"EVERYONE", true, true},
		{"garbage", false, false},
	}
	for _, c := range cases {
		if got := canBotMessage(c.privacy, c.follows); got != c.want {
			t.Errorf("canBotMessage(%q, %v) = %v, want %v", c.privacy, c.follows, got, c.want)
		}
	}
}

func TestDecideInitiation(t *testing.T) {
	now := time.Now()
	bot := "bot1"
	silence := 3 * 24 * time.Hour

	// No conversation -> opener.
	if d := decideInitiation(bot, now, silence, nil); d != "opener" {
		t.Errorf("nil messages => %q, want opener", d)
	}
	if d := decideInitiation(bot, now, silence, []policyMessage{}); d != "opener" {
		t.Errorf("empty messages => %q, want opener", d)
	}

	// Human has replied -> skip.
	msgs := []policyMessage{
		{SenderID: "bot1", CreatedAt: now.Add(-10 * time.Hour)},
		{SenderID: "human1", CreatedAt: now.Add(-9 * time.Hour)},
	}
	if d := decideInitiation(bot, now, silence, msgs); d != "skip" {
		t.Errorf("human replied => %q, want skip", d)
	}

	// Two unanswered bot messages -> skip.
	two := []policyMessage{
		{SenderID: "bot1", CreatedAt: now.Add(-10 * time.Hour)},
		{SenderID: "bot1", CreatedAt: now.Add(-9 * time.Hour)},
	}
	if d := decideInitiation(bot, now, silence, two); d != "skip" {
		t.Errorf("two bot msgs => %q, want skip", d)
	}

	// One unanswered bot opener, enough silence -> followup.
	old := []policyMessage{{SenderID: "bot1", CreatedAt: now.Add(-4 * 24 * time.Hour)}}
	if d := decideInitiation(bot, now, silence, old); d != "followup" {
		t.Errorf("one old bot msg => %q, want followup", d)
	}

	// One unanswered bot opener, not enough silence -> skip.
	recent := []policyMessage{{SenderID: "bot1", CreatedAt: now.Add(-1 * time.Hour)}}
	if d := decideInitiation(bot, now, silence, recent); d != "skip" {
		t.Errorf("one recent bot msg => %q, want skip", d)
	}
}

func TestFormatDmHistory(t *testing.T) {
	msgs := []dmMessage{
		{SenderID: "human1", Content: "hi"},
		{SenderID: "bot1", Content: "hello"},
	}
	turns := formatDmHistory(msgs, "bot1")
	if len(turns) != 2 {
		t.Fatalf("got %d turns", len(turns))
	}
	if turns[0].From != "them" || turns[0].Text != "hi" {
		t.Errorf("turn0 = %+v, want them/hi", turns[0])
	}
	if turns[1].From != "you" || turns[1].Text != "hello" {
		t.Errorf("turn1 = %+v, want you/hello", turns[1])
	}
}

func TestOrderPair(t *testing.T) {
	a, b := orderPair("zeta", "alpha")
	if a != "alpha" || b != "zeta" {
		t.Errorf("orderPair = (%q,%q), want (alpha,zeta)", a, b)
	}
	a, b = orderPair("alpha", "zeta")
	if a != "alpha" || b != "zeta" {
		t.Errorf("orderPair = (%q,%q), want (alpha,zeta)", a, b)
	}
}
